File: ios/Framework/Core/Auth/UserPortraitScreenlet/UserPortraitScreenlet.swift
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */
import UIKit


@objc public protocol UserPortraitScreenletDelegate : BaseScreenletDelegate {

	/// Called when an image is received from the server. You can then apply image filters 
	/// (grayscale, for example) and return the new image. You can return the original image 
	/// supplied as the argument if you don’t want to modify it.
	///
	/// - Parameters:
	///   - screenlet
	///   - image: user portrait image.
	/// - Returns: original or modify image.
	@objc optional func screenlet(_ screenlet: UserPortraitScreenlet,
			onUserPortraitResponseImage image: UIImage) -> UIImage

	/// Called when an error occurs in the process.
	/// The NSError object describes the error.
	///
	/// - Parameters:
	///   - screenlet
	///   - error: error while retrieving user portrait image.
	@objc optional func screenlet(_ screenlet: UserPortraitScreenlet,
			onUserPortraitError error: NSError)

	/// Called when a new portrait is uploaded to the server. You receive the user 
	/// attributes as a parameter.
	///
	/// - Parameters:
	///   - screenlet
	///   - attributes: user portrait attributes.
	@objc optional func screenlet(_ screenlet: UserPortraitScreenlet,
			onUserPortraitUploaded attributes: [String:AnyObject])

	/// Called when an error occurs in the upload process.
	/// The NSError object describes the error.
	///
	/// - Parameters:
	///   - screenlet
	///   - error: error while uploading the user portrait image.
	@objc optional func screenlet(_ screenlet: UserPortraitScreenlet,
			onUserPortraitUploadError error: NSError)
}


open class UserPortraitScreenlet: BaseScreenlet {


	//MARK: Inspectables

	@IBInspectable open var borderWidth: CGFloat = 1.0 {
		didSet {
			(screenletView as? UserPortraitViewModel)?.borderWidth = self.borderWidth
		}
	}

	@IBInspectable open var borderColor: UIColor? {
		didSet {
			(screenletView as? UserPortraitViewModel)?.borderColor = self.borderColor
		}
	}

	@IBInspectable open var editable: Bool = false {
		didSet {
			screenletView?.editable = self.editable
		}
	}

	@IBInspectable open var offlinePolicy: String? = CacheStrategyType.remoteFirst.rawValue


	open var userPortraitDelegate: UserPortraitScreenletDelegate? {
		return self.delegate as? UserPortraitScreenletDelegate
	}

	open var viewModel: UserPortraitViewModel {
		return screenletView as! UserPortraitViewModel
	}

	open var userId: Int64? {
		return loadedUserId
	}

	fileprivate var loadedUserId: Int64?


	//MARK: BaseScreenlet

	override open func onCreated() {
		super.onCreated()

		viewModel.borderWidth = self.borderWidth
		viewModel.borderColor = self.borderColor
		screenletView?.editable = self.editable
	}

	override open func createInteractor(name: String, sender: AnyObject?) -> Interactor? {
		let interactor: Interactor?

		if isActionRunning(name) {
			cancelInteractorsForAction(name)
		}

		switch name {
		case "load-portrait":
			let loadInteractor = sender as! DownloadUserPortraitInteractor
			interactor = loadInteractor

			loadInteractor.cacheStrategy = CacheStrategyType(rawValue: self.offlinePolicy ?? "") ?? .remoteFirst

			loadInteractor.onSuccess = {
				if loadInteractor.userHasDefaultPortrait, let user = loadInteractor.resultUser {
					self.viewModel.loadPlaceholder(for: user)
				}
				else if let imageValue = loadInteractor.resultImage {
					let finalImage = self.userPortraitDelegate?.screenlet?(self, onUserPortraitResponseImage: imageValue)

					self.loadedUserId = loadInteractor.resultUser?.userId ?? self.loadedUserId
					self.setPortraitImage(finalImage ?? imageValue)
				}
				else {
					self.userPortraitDelegate?.screenlet?(self,
					                                      onUserPortraitError: NSError.errorWithCause(.invalidServerResponse,
															message: "Could not load user portrait image."))

					self.loadedUserId = nil
					self.setPortraitImage(nil)
				}
			}

			loadInteractor.onFailure = {
				self.userPortraitDelegate?.screenlet?(self, onUserPortraitError: $0)

				self.loadedUserId = nil
				self.setPortraitImage(nil)
			}

		case "upload-portrait":
			let image = sender as! UIImage
			let userId: Int64

			if let loadedUserIdValue = loadedUserId {
				userId = loadedUserIdValue
			}
			else {
				print("ERROR: Can't change the portrait without an userId\n")

				return nil
			}

			let uploadInteractor = UploadUserPortraitInteractor(
				screenlet: self,
				userId: userId,
				image: image)
			interactor = uploadInteractor

			uploadInteractor.cacheStrategy = CacheStrategyType(rawValue: self.offlinePolicy ?? "") ?? .remoteFirst

			uploadInteractor.onSuccess = {
				self.userPortraitDelegate?.screenlet?(self, onUserPortraitUploaded: uploadInteractor.uploadResult!)

				self.loadedUserId = uploadInteractor.userId
				self.setPortraitImage(uploadInteractor.image)
			}

			uploadInteractor.onFailure = {
				self.userPortraitDelegate?.screenlet?(self, onUserPortraitUploadError: $0)
			}
			
		default:
			interactor = nil
		}
		
		return interactor
	}


	//MARK: Public methods

	/// Loads the user portrait that correspond to the user logged.
	///
	/// - Returns: true if suceed, false otherwise.
	open func loadLoggedUserPortrait() -> Bool {
		guard let userId = SessionContext.currentContext?.user.userId else {
			return false
		}

		let interactor = DownloadUserPortraitInteractor(
			screenlet: self,
			userId: userId)

		loadedUserId = userId

		return performAction(name: "load-portrait", sender: interactor)
	}

	/// Loads the user portrait image.
	///
	/// - Parameters:
	///   - portraitId: portrait identifier.
	///   - uuid: user portrait unique identifier.
	///   - male: true if the user is male, false otherwise.
	/// - Returns: true if succeed, false otherwise.
	@discardableResult
	open func load(portraitId: Int64, uuid: String, male: Bool = true) -> Bool {
		let interactor = DownloadUserPortraitInteractor(
				screenlet: self,
				portraitId: portraitId,
				uuid: uuid,
				male: male)

		loadedUserId = nil

		return performAction(name: "load-portrait", sender: interactor)
	}

	/// Loads the user portrait image throught user identifier.
	///
	/// - Parameter userId: user identifier.
	/// - Returns: true if succeed, false otherwise.
	@discardableResult
	open func load(userId: Int64) -> Bool {
		let interactor = DownloadUserPortraitInteractor(
				screenlet: self,
				userId: userId)

		loadedUserId = userId

		return performAction(name: "load-portrait", sender: interactor)
	}

	/// Loads the user portrait throught company identifier and email address.
	///
	/// - Parameters:
	///   - companyId: company identifier.
	///   - emailAddress: user email.
	/// - Returns: true if succeed, false otherwise.
	@discardableResult
	open func load(companyId: Int64, emailAddress: String) -> Bool {
		let interactor = DownloadUserPortraitInteractor(
				screenlet: self,
				companyId: companyId,
				emailAddress: emailAddress)

		loadedUserId = nil

		return performAction(name: "load-portrait", sender: interactor)
	}

	/// Loads the user portrait throught company identifier and user screen name.
	///
	/// - Parameters:
	///   - companyId: company identifier.
	///   - screenName: user screen name.
	/// - Returns: <#return value description#>
	@discardableResult
	open func load(companyId: Int64, screenName: String) -> Bool {
		let interactor = DownloadUserPortraitInteractor(
				screenlet: self,
				companyId: companyId,
				screenName: screenName)

		loadedUserId = nil

		return performAction(name: "load-portrait", sender: interactor)
	}

	/// Loads the user portrait placeholder when the user portrait is empty.
	open func loadPlaceholder() {
		setPortraitImage(nil)
	}


	//MARK: Private methods

	fileprivate func setPortraitImage(_ image: UIImage?) {
		viewModel.image = image

		if image == nil {
			let error = NSError.errorWithCause(.abortedDueToPreconditions,
					message: "Could not set user portrait image.")
			userPortraitDelegate?.screenlet?(self, onUserPortraitError: error)
		}
	}

}

File: android/library/src/main/java/com/liferay/mobile/screens/userportrait/UserPortraitScreenlet.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 * <p/>
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * <p/>
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.mobile.screens.userportrait;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import com.liferay.mobile.screens.R;
import com.liferay.mobile.screens.base.BaseScreenlet;
import com.liferay.mobile.screens.base.MediaStoreRequestShadowActivity;
import com.liferay.mobile.screens.base.MediaStoreRequestShadowActivity.MediaStoreCallback;
import com.liferay.mobile.screens.base.interactor.Interactor;
import com.liferay.mobile.screens.context.LiferayScreensContext;
import com.liferay.mobile.screens.context.SessionContext;
import com.liferay.mobile.screens.context.User;
import com.liferay.mobile.screens.userportrait.interactor.UserPortraitInteractorListener;
import com.liferay.mobile.screens.userportrait.interactor.load.UserPortraitLoadInteractor;
import com.liferay.mobile.screens.userportrait.interactor.upload.UserPortraitUploadEvent;
import com.liferay.mobile.screens.userportrait.interactor.upload.UserPortraitUploadInteractor;
import com.liferay.mobile.screens.userportrait.view.UserPortraitViewModel;
import com.liferay.mobile.screens.util.LiferayLogger;

/**
 * @author Javier Gamarra
 * @author Jose Manuel Navarro
 */
public class UserPortraitScreenlet extends BaseScreenlet<UserPortraitViewModel, Interactor>
    implements UserPortraitInteractorListener {

    public static final String UPLOAD_PORTRAIT = "UPLOAD_PORTRAIT";
    public static final String LOAD_PORTRAIT = "LOAD_PORTRAIT";
    public static final String LOAD_CURRENT_USER = "LOAD_CURRENT_USER";
    private static final String STATE_SUPER = "userportrait-super";
    private boolean autoLoad;
    private boolean male;
    private long portraitId;
    private long userId;
    private String uuid;
    private boolean editable;
    private UserPortraitListener listener;

    public UserPortraitScreenlet(Context context) {
        super(context);
    }

    public UserPortraitScreenlet(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public UserPortraitScreenlet(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public UserPortraitScreenlet(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    /**
     * Searches the user portrait with the given attributes ({@link #portraitId}
     * and {@link #uuid} or {@link #userId}) and loads it in the screenlet.
     */
    public void load() {
        performUserAction(LOAD_PORTRAIT);
    }

    /**
     * Loads the user portrait that correspond to the user logged.
     */
    public void loadLoggedUserPortrait() {
        performUserAction(LOAD_CURRENT_USER);
    }

    /**
     * Selects a new user portrait from the device camera.
     */
    public void openCamera() {
        MediaStoreRequestShadowActivity.show(getContext(), MediaStoreRequestShadowActivity.TAKE_PICTURE_WITH_CAMERA,
            new MediaStoreCallback() {

                @Override
                public void onUriReceived(Uri uri) {
                    onPictureUriReceived(uri);
                }
            });
    }

    /**
     * Selects a new user portrait from the device image gallery.
     */
    public void openGallery() {
        MediaStoreRequestShadowActivity.show(getContext(), MediaStoreRequestShadowActivity.SELECT_IMAGE_FROM_GALLERY,
            new MediaStoreCallback() {

                @Override
                public void onUriReceived(Uri uri) {
                    onPictureUriReceived(uri);
                }
            });
    }

    @Override
    public Bitmap onEndUserPortraitLoadRequest(Bitmap bitmap) {
        Bitmap finalImage = bitmap;

        if (listener != null) {
            finalImage = listener.onUserPortraitLoadReceived(bitmap);

            if (finalImage == null) {
                finalImage = bitmap;
            }
        }

        getViewModel().showFinishOperation(finalImage);

        return finalImage;
    }

    @Override
    public void onUserPortraitUploaded(Long uuid) {
        if (listener != null) {
            listener.onUserPortraitUploaded();
        }

        getViewModel().showFinishOperation(UPLOAD_PORTRAIT);

        ((UserPortraitLoadInteractor) getInteractor(LOAD_PORTRAIT)).start(uuid);
    }

    public void onPictureUriReceived(Uri pictureUri) {
        performUserAction(UPLOAD_PORTRAIT, pictureUri);
    }

    @Override
    public void onUserWithoutPortrait(User user) {
        getViewModel().showPlaceholder(user);
        getViewModel().showFinishOperation(LOAD_PORTRAIT);
    }

    @Override
    public void error(Exception e, String userAction) {
        if (listener != null) {
            listener.error(e, userAction);
        }

        getViewModel().showFailedOperation(userAction, e);
    }

    public UserPortraitListener getListener() {
        return listener;
    }

    public void setListener(UserPortraitListener listener) {
        this.listener = listener;
    }

    public boolean isMale() {
        return male;
    }

    public void setMale(boolean male) {
        this.male = male;
    }

    public long getPortraitId() {
        return portraitId;
    }

    public void setPortraitId(long portraitId) {
        this.portraitId = portraitId;
    }

    public String getUuid() {
        return uuid;
    }

    public void setUuid(String uuid) {
        this.uuid = uuid;
    }

    public boolean getEditable() {
        return editable;
    }

    public boolean isAutoLoad() {
        return autoLoad;
    }

    public void setAutoLoad(boolean autoLoad) {
        this.autoLoad = autoLoad;
    }

    public boolean isEditable() {
        return editable;
    }

    public void setEditable(boolean editable) {
        this.editable = editable;
    }

    @Override
    public long getUserId() {
        return userId;
    }

    @Override
    public void setUserId(long userId) {
        this.userId = userId;
    }

    /**
     * Checks if exists {@link #portraitId} and {@link #uuid} or {@link #userId} attributes
     * and then calls {@link #load()} method. It's not necessary to be logged in.
     */
    protected void autoLoad() {
        if (((portraitId != 0) && (uuid != null)) || (userId != 0)) {
            try {
                load();
            } catch (Exception e) {
                LiferayLogger.e("Error loading portrait", e);
            }
        }
    }

    @Override
    protected View createScreenletView(Context context, AttributeSet attributes) {
        TypedArray typedArray =
            context.getTheme().obtainStyledAttributes(attributes, R.styleable.UserPortraitScreenlet, 0, 0);

        autoLoad = typedArray.getBoolean(R.styleable.UserPortraitScreenlet_autoLoad, true);
        male = typedArray.getBoolean(R.styleable.UserPortraitScreenlet_male, true);
        portraitId = typedArray.getInt(R.styleable.UserPortraitScreenlet_portraitId, 0);
        uuid = typedArray.getString(R.styleable.UserPortraitScreenlet_uuid);
        editable = typedArray.getBoolean(R.styleable.UserPortraitScreenlet_editable, false);
        userId = typedArray.getInt(R.styleable.UserPortraitScreenlet_userId, 0);

        int layoutId = typedArray.getResourceId(R.styleable.UserPortraitScreenlet_layoutId, getDefaultLayoutId());

        typedArray.recycle();

        return LayoutInflater.from(context).inflate(layoutId, null);
    }

    @Override
    protected Interactor createInteractor(String actionName) {
        if (UPLOAD_PORTRAIT.equals(actionName)) {
            return new UserPortraitUploadInteractor();
        } else {
            return new UserPortraitLoadInteractor();
        }
    }

    @Override
    protected void onUserAction(String userActionName, Interactor interactor, Object... args) {
        switch (userActionName) {
            case UPLOAD_PORTRAIT:
                UserPortraitUploadInteractor userPortraitInteractor =
                    (UserPortraitUploadInteractor) getInteractor(userActionName);
                Uri path = (Uri) args[0];
                if (userId != 0) {
                    userPortraitInteractor.start(new UserPortraitUploadEvent(path));
                }
                break;
            case LOAD_CURRENT_USER:
                UserPortraitLoadInteractor userPortraitLoadInteractorCurrentUser =
                    (UserPortraitLoadInteractor) getInteractor(userActionName);
                long currentUserId = SessionContext.isLoggedIn() ? SessionContext.getCurrentUser().getId() : 0;
                userPortraitLoadInteractorCurrentUser.start(currentUserId);
                break;
            default:
                UserPortraitLoadInteractor userPortraitLoadInteractor =
                    (UserPortraitLoadInteractor) getInteractor(userActionName);
                if (portraitId != 0 && uuid != null) {
                    userPortraitLoadInteractor.start(male, portraitId, uuid);
                } else {
                    userPortraitLoadInteractor.start(userId);
                }
                break;
        }
    }

    @Override
    protected void onScreenletAttached() {
        if (autoLoad) {
            autoLoad();
        }
    }

    @Override
    protected void onRestoreInstanceState(Parcelable inState) {
        Bundle bundle = (Bundle) inState;
        super.onRestoreInstanceState(bundle.getParcelable(STATE_SUPER));
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Bundle bundle = new Bundle();
        bundle.putParcelable(STATE_SUPER, super.onSaveInstanceState());
        return bundle;
    }

    private void startShadowActivityForMediaStore(int mediaStore) {
        Activity activity = LiferayScreensContext.getActivityFromContext(getContext());

        Intent intent = new Intent(activity, MediaStoreRequestShadowActivity.class);
        intent.putExtra(MediaStoreRequestShadowActivity.MEDIA_STORE_TYPE, mediaStore);

        activity.startActivity(intent);
    }
}